### PR TITLE
Only create per-object webhooks for configured types

### PR DESCRIPTION
**Pull request by erikgb**
Fixes https://github.com/kubernetes-sigs/hierarchical-namespaces/issues/152

### CODE REVIEW

**Comment by erikgb** [comment]
@adrianludwin Existing tests are running fine now. Before writing some tests for this new feature, I would like to get your opinion on the approach. Because of the cert-rotator (as always 😉), I cannot move the now dynamic objects-webhook to a dedicated validating webhook configuration resource. IMO we should not mess with resources provisioned as part of HNC. That will never work with modern GitOps-tools like FluxCD/ArgoCD - as their reconcilers will compete with our internal reconciler. It "might" work if it was possible to create a webhook without rules (for the initial configuration), but it seems like this is not allowed by the API server.

**Comment by adrianludwin** [comment]
> @adrianludwin Existing tests are running fine now. Before writing some tests for this new feature, I would like to get your opinion on the approach. Because of the cert-rotator (as always 😉), I cannot move the now dynamic objects-webhook to a dedicated validating webhook configuration resource. IMO we should not mess with resources provisioned as part of HNC. That will never work with modern GitOps-tools like FluxCD/ArgoCD - as their reconcilers will compete with our internal reconciler.

Is this only a problem with cert-rotator? Or would cert-manager have the same problem? I would have thought that _any_ controller that tried to inject certs into the webhook would have the same problem?

How does cert-manager work with gitops tools today? Is there no way to say "please ignore these fields?"

> This was not possible when defining the webhook with the kubebuilder markers, but I worked around this by extending the patch to create the complete webhook (without rules).

Does this have something to do with the cert-rotator problem? I thought the caBundle needs to be injected to every new webhook (e.g. there isn't one caBundle per config) so why does moving these to a new config help? Just so Flux/Argo don't try to reconcile it away?

Instead of hand-writing a patch, why not create the _entire_ webhook config in code? There's not a lot in a webhook config other than the list of webhooks.

**Comment by erikgb** [comment]
> > @adrianludwin Existing tests are running fine now. Before writing some tests for this new feature, I would like to get your opinion on the approach. Because of the cert-rotator (as always 😉), I cannot move the now dynamic objects-webhook to a dedicated validating webhook configuration resource. IMO we should not mess with resources provisioned as part of HNC. That will never work with modern GitOps-tools like FluxCD/ArgoCD - as their reconcilers will compete with our internal reconciler.
> 
> Is this only a problem with cert-rotator? Or would cert-manager have the same problem? I would have thought that _any_ controller that tried to inject certs into the webhook would have the same problem?

I think my formulation was bad, sorry. The real problem is not with cert-rotator (or cert-manager). But with cert-manager, we could create a dedicated dynamic `ValidatingWebhookConfiguration` from code, and not mess with the VWC statically provisioned as part of the HNC deployment. Having multiple controllers working on the same resource is error-prone, but as long as the controllers do not touch the same fields, it can work. So without specifying the object-webhook rules in the static HNC configuration, this seems to work ok.

> > This was not possible when defining the webhook with the kubebuilder markers, but I worked around this by extending the patch to create the complete webhook (without rules).
> 
> Does this have something to do with the cert-rotator problem? I thought the caBundle needs to be injected to every new webhook (e.g. there isn't one caBundle per config) so why does moving these to a new config help? Just so Flux/Argo don't try to reconcile it away?
> 
> Instead of hand-writing a patch, why not create the _entire_ webhook config in code? There's not a lot in a webhook config other than the list of webhooks.

WDYM? The kubebuilder markers does not allow a user to configure a webhook without rules. That is why I suggest to add the complete object-webhook using a patch. All other webhooks are sourced from kubebuilder-markers, so the `webhook/manifests.yaml` file is generated.

**Comment by adrianludwin** [comment]
... I guess the only problem with creating it automatically is that if users want to disable it somehow, they can't. Maybe that's a good reason to have HNC rely on the VWC config existing:

- Users can delete _all_ object webhooks simply by deleting the config, and HNC won't recreate it
- We could let users add annotations to the VWC (e.g. "don't update" for temporary investigations). We don't need this on day 1, but it's a nice future option.

Given this, a hand-written, empty VWC sgtm!

**Comment by erikgb** [comment]
@adrianludwin I think the code is almost done and ready for review. All existing tests are passing, and I think this indirectly verifies that this feature works. The e2e-tests would never succeed if the objects webhook is not configured correctly.
But I wonder if we should test some more corner cases in the integration tests? Writing unit tests for the new code seems almost impossible if I am supposed to follow the existing code style - which I think you prefer. That would eventually require more types to be exported as tests of reconcilers are currently located in a dedicated test-package.  WDYT?

**Comment by mzain** [comment]
Hey @adrianludwin @erikgb when do we think this will be merged and available for use? 
it is an important feature for our usage of HNC.

**Comment by erikgb** [comment]
@mzain We have sadly abandoned HNC because it turned out to be impossible to run it safely in our clusters. We are currently evaluating a simpler operator, https://github.com/cybozu-go/accurate - that has exactly the features we need. I think @adrianludwin is quite busy with other stuff, so HNC seems like it's dormant at present.

I am not going to follow up on this PR, so feel free to pick it up! 😸

**Comment by adrianludwin** [comment]
Sorry to hear @erikgb , yup just been too busy to follow up :(

**Comment by mzain** [comment]
Just curious @adrianludwin
What's required in this PR, seems pretty complete? Since it's an important feature for us, we may able to pick up the PR and get it merged.

Also a second unrelated question, I can't see any roadmap or a defined cadence for releases. Is that ad-hoc? If yes, when do you think we would get the new release with this change?

**Review by adrianludwin — COMMENTED** [review]
Sorry it's taken me so long to get to this. @erikgb if you're able to answer some questions I'd appreciate it even if you can't work on this yourself anymore.

I think it's basically ready to go in, and we don't need to test webhooks in the integ tests (as you say). I think that a bit of manual testing, and verifying that there's at least one e2e test that would break if this were broken, is probably good enough.

@mzaian if you're able to pick this up that would be great. Otherwise @rjbez17 I'd appreciate it if you have an hour or so to check this out too!

File: internal/integtest/setup.go [hunk 1]
```diff
@@ -69,8 +76,28 @@ func HNCBeforeSuite() {
 	SetDefaultEventuallyTimeout(time.Second * 4)
 
 	By("configuring test environment")
+	sideEffectClassNone := apiadmissionregistrationv1.SideEffectClassNone
 	testEnv = &envtest.Environment{
 		CRDDirectoryPaths: []string{filepath.Join("..", "..", "config", "crd", "bases")},
+		WebhookInstallOptions: envtest.WebhookInstallOptions{
```
[the remark this answers is not in the document]
Comment (erikgb) [reply internal/integtest/setup.go:82]: I don't remember, but there was probably a reason. I suggest attempting to remove it and see if the tests passes.

File: internal/integtest/setup.go [hunk 2]
```diff
@@ -94,13 +121,20 @@ func HNCBeforeSuite() {
...
 		NewClient:          config.NewClient(false),
 		MetricsBindAddress: "0", // disable metrics serving since 'go test' runs multiple suites in parallel processes
 		Scheme:             scheme.Scheme,
+		Host:               webhookInstallOptions.LocalServingHost,
+		Port:               webhookInstallOptions.LocalServingPort,
+		CertDir:            webhookInstallOptions.LocalServingCertDir,
 	})
 	Expect(err).ToNot(HaveOccurred())
 
+	// Register a dummy webhook since the test control plane is to test reconcilers
```
[the remark this answers is not in the document]
Comment (erikgb) [reply internal/integtest/setup.go:135]: Same answer as before. I think the tests are failing without it. But worth a try.

**Comment by rjbez17** [comment]
@erikgb did you want to finish fixing the last few items here and we can get it merged in?

**Comment by erikgb** [comment]
@rjbez17 it seems to me like it's more to do than minor adjustments to make @adrianludwin happy. 😊 Feel free to pick up this PR.

File: config/webhook/webhook_patch.yaml [hunk 3]
```diff
@@ -19,19 +28,5 @@ webhooks:
...
-    - '*'
-    operations:
-    - CREATE
-    - UPDATE
-    - DELETE
-    resources:
-    - '*'
-    scope: "Namespaced"
+  # Intentionally no rules, as these are maintained by the HNCConfiguration reconciler.
+  # At present controller-gen is unable to generate a webhook without rules from kubebuilder markers.
```
[the remark this answers is not in the document]
Comment (erikgb) [reply config/webhook/webhook_patch.yaml:32]: To the top? Do you mean this literarly? Top of the file?

[the remark this answers is not in the document]
Comment (erikgb) [reply config/webhook/webhook_patch.yaml]: Moved the comment to the top of the file.

**Comment by erikgb** [comment]
As expected, it fails now: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_hierarchical-namespaces/285/pull-hnc-test/1766783682422509568. If I remember this correctly, a webhook is required in the reconciler integration tests because the reconciler is now managing a webhook. And since the envTest represents a real control-plane - we also need to mock/fake the webhook endpoint. I am going to revert the revert now, and hope you can look at this PR again - with the understanding that the new webhook setup IS required in the reconciler integration test.
